crates/wysiwyg/src/composer_model/replace_text.rs

```diff
@@ -179,41 +184,100 @@ where
             }
         }
 
+        // If there was a previous line break, we need to split the code block and add the line break
         if has_previous_line_break {
-            let DomNode::Container(sub_tree) = self.state.dom.split_sub_tree(&leaf.node_handle, leaf.start_offset - selection_offset, block_location.node_handle.depth()) else {
+            let mut sub_tree = self.state.dom.split_sub_tree_from(
+                &leaf.node_handle,
+                leaf.start_offset - selection_offset,
+                block_location.node_handle.depth(),
+            );
+            sub_tree.set_handle(DomHandle::root());
+            let DomNode::Container(sub_tree_container) = &mut sub_tree else {
                 panic!("Sub tree must start from a container node");
             };
-            let DomNode::Container(block) = self.state.dom.lookup_node_mut(&block_location.node_handle) else {
-                panic!("Parent must be a block node");
-            };
-            if block.children().is_empty() {
-                self.state.dom.replace(
-                    &block_location.node_handle,
-                    vec![
-                        DomNode::new_line_break(),
-                        DomNode::new_code_block(sub_tree.children().clone()),
-                    ],
-                );
-            } else {
-                let next_handle = block_location.node_handle.next_sibling();
-                if !sub_tree.children().is_empty() {
-                    self.state.dom.insert_at(
-                        &next_handle,
-                        DomNode::new_code_block(sub_tree.children().clone()),
+
+            let new_line_at_end = leaf.is_start();
+            fn insert_code_block<S: UnicodeString>(
+                dom: &mut Dom<S>,
+                sub_tree: DomNode<S>,
+                at_handle: &DomHandle,
+                new_line_at_end: bool,
+            ) {
+                dom.insert_at(&at_handle, sub_tree);
+                let insert_new_line_at = if new_line_at_end {
+                    at_handle.next_sibling()
+                } else {
+                    at_handle.clone()
+                };
+                dom.insert_at(&insert_new_line_at, DomNode::new_line_break());
+            }
```

Comment:
Why not having this function as part of file instead of being inside the scope of another function?

Reply:
It's internal to the function because it's only a helper for this function (we have to use it twice). I don't think it's worth it to have an `insert_X_node_at()` function for a concrete type.

Reply:
I see... could we at least place it either at the end of the enclosing scope or at the start of the enclosing scope?
Since having it in the middle of the enclosing scope kinda "breaks" it.

Reply:
Yes, it's true that it looks weird at its current position 😅 . Done in ed191dc.